### PR TITLE
Feature/change from jsx to tsx breadcrumbs and footer

### PR DESCRIPTION
**Pull request by remibengtbouvet**

### CODE REVIEW

File: src/molecules/Breadcrumbs/Breadcrumbs.pcss [hunk 1]
```diff
@@ -1,9 +1,12 @@
 .telia-breadcrumbs {
+  padding-top: 8px;
+  padding-bottom: 8px;
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.pcss:3]: Seeing as the scenario  of having breadcrmubs on some pages just belo main menu
- 4px was slightly not enough as "Main Menu" on telia has a 4px shadow box...
-

File: src/molecules/Breadcrumbs/Breadcrumbs.pcss [hunk 2]
```diff
@@ -1,9 +1,12 @@
...
+
   &--core-purple {
     background-color: var(--core-purple);
   }
 
-  &--gray {
+  &--grey {
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.pcss:9]: Ive moved it to grey - as we use EN not US

File: src/molecules/Breadcrumbs/Breadcrumbs.pcss [hunk 3]
```diff
@@ -50,8 +52,8 @@
       color: var(--black) !important;
     }
 
-    &--gray {
-      color: var(--gray-200) !important;
+    &--grey {
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.pcss:55]: EN not US

File: src/molecules/Breadcrumbs/Breadcrumbs.pcss [hunk 4]
```diff
@@ -68,8 +70,8 @@
...
+      color: var(--grey-200) !important;
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.pcss:74]: EN not US 
gray-200 does not exist anyways

File: src/molecules/Breadcrumbs/Breadcrumbs.pcss [hunk 5]
```diff
@@ -102,8 +104,8 @@
...
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.pcss:108]: ....

File: src/molecules/Breadcrumbs/Breadcrumbs.pcss [hunk 6]
```diff
@@ -119,8 +121,8 @@
...
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.pcss:125]: ..

File: src/molecules/Breadcrumbs/Breadcrumbs.pcss [hunk 7]
```diff
@@ -61,22 +67,25 @@
     padding: 0.6rem !important;
 
     &--white {
-      color: var(--white) !important;
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.pcss:64]: Always try to code without important

File: src/molecules/Breadcrumbs/Breadcrumbs.pcss [hunk 8]
```diff
@@ -61,22 +67,25 @@
...
+    }
+
+    &--grey {
+      color: var(--grey-200);
     }
 
-    &--gray {
-      color: var(--gray-200) !important;
+    &--core-purple {
+      color: var(--core-purple);
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.pcss:82]: Added option for core-pjurple

File: src/molecules/Breadcrumbs/Breadcrumbs.pcss [hunk 9]
```diff
@@ -43,15 +45,19 @@
...
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.pcss:60]: Added option for core-pjurple

File: src/molecules/Breadcrumbs/Breadcrumbs.pcss [hunk 10]
```diff
@@ -61,22 +67,25 @@
...
     }
 
     &:hover,
     &:visited,
     &:active,
     &:link {
-      color: inherit;
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.pcss:79]: color:inherit? ... 
Prolly should never use the word "inherit" in CSS

File: src/molecules/Breadcrumbs/Breadcrumbs.stories.tsx [hunk 11]
```diff
@@ -7,39 +7,196 @@ export default {
...
+    <p>{'Remember: Mobile view will maximum display two breadcrumbs, controlled through CSS'}</p>
+    <Breadcrumbs
+      crumbs={breadCrumbsData.filter((item, i) => i <= 0)}
+      alwaysShowRootCrumb={true}
+      pageSize={2}
+      pagingSize={2}
+      fontColor={''}
+      iconColor={''}
+      backgroundColor={''}
+    />
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.stories.tsx:20]: Good - me like

File: src/molecules/Breadcrumbs/Breadcrumbs.tsx [hunk 12]
```diff
@@ -51,32 +77,48 @@ const Breadcrumbs = (props) => {
...
 
-  const rangeVisibleCrumbs = (start, end) => {
+  const rangeVisibleCrumbs = (start: number, end: number) => {
     return Array(end - start + 1)
-      .fill()
+      .fill(null)
       .map((_, idx) => start + idx);
   };
-  var result = rangeVisibleCrumbs(minIndex, maxIndex - 1);
+  const result = rangeVisibleCrumbs(minIndex, maxIndex - 1);
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.tsx:97]: result? 📦 
ranges?

indexRanges?
keys?

File: src/molecules/Breadcrumbs/Breadcrumbs.tsx [hunk 13]
```diff
@@ -51,32 +77,48 @@ const Breadcrumbs = (props) => {
...
-  const getStyle = (style, color) => {
+  const getStyle = (style: string, color: string) => {
     if (color) {
       return (style += ' ' + style + '--' + color);
     }
     return style;
   };
 
-  const rangeVisibleCrumbs = (start, end) => {
+  const rangeVisibleCrumbs = (start: number, end: number) => {
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.tsx:92]: try to always prefix a method which returns something with:
get
has
is
...
verb

File: src/molecules/Breadcrumbs/Breadcrumbs.tsx [hunk 14]
```diff
@@ -128,52 +170,80 @@ const Breadcrumbs = (props) => {
     setMaxIndex(Math.min(maxIndex + pagingSize, maxCrumbIndex));
   };
 
-  const CrumbLabel = ({ crumb }) => {
+  const CrumbLabel = (props: { fontColor: string; name: string | undefined }) => {
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.tsx:173]: Why have you done this?

This is ugly, or do you really like this?

You really want to send an object, and object holds the params/varaibles. It is so much cleaner - less words.

File: src/molecules/Breadcrumbs/Breadcrumbs.tsx [hunk 15]
```diff
@@ -128,52 +170,80 @@ const Breadcrumbs = (props) => {
...
     return (
       <span className={getStyle('telia-breadcrumbs__crumb-label', props.fontColor)}>
-        <b>{crumb.name}</b>
+        <b>{props.name}</b>
       </span>
     );
   };
 
-  const CrumbPaging = ({ onPagingEvent }) => {
+  const CrumbPaging = (props: {
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.tsx:181]: Why convert to "props" everywhere?

props is a "react thing" - at least the "props" meaning when  Iread it in a JS/TS context - I read it as some properties belongs within a state in some parent above where I currently am.

File: src/molecules/Breadcrumbs/Breadcrumbs.tsx [hunk 16]
```diff
@@ -128,52 +170,80 @@ const Breadcrumbs = (props) => {
...
-
-        {crumb.type === crumbType.LABEL && <CrumbLabel crumb={crumb} />}
-
-        {crumb.type === crumbType.RIGHT && <CrumbPaging onPagingEvent={onPagingRight} />}
-
-        {crumb.type === crumbType.LINK && <CrumbLink crumb={crumb} />}
+      <li key={props.key} className={'telia-breadcrumbs__crumb'}>
+        {props.type === crumbType.LEFT && <CrumbPaging onPagingEvent={onPagingLeft} iconColor={props.iconColor} />}
+        {props.type === crumbType.LABEL && <CrumbLabel fontColor={props.fontColor} name={props.name} />}
+        {props.type === crumbType.RIGHT && <CrumbPaging onPagingEvent={onPagingRight} iconColor={props.iconColor} />}
```
Comment (oyvindbergbouvet) [inline src/molecules/Breadcrumbs/Breadcrumbs.tsx:235]: Please add new empty lines between each of these. As it was.

AIr/space is really neat in code.

OIts like our breadcrumb, i show you an image - slightly more padding on the breadcrumb "frees" the breadcrumb? And everything wants to be free? :P

File: src/molecules/Footer/Footer.tsx [hunk 17]
```diff
@@ -120,15 +169,15 @@ const Footer = (props) => {
...
+  const PrivacyAndCookies = (props: { color: boolean; url: string; name: string }) => {
     return (
       <a className={'telia-footer__white' + setColor(props.color)} href={props.url}>
         {props.name}
       </a>
     );
   };
 
-  const Facebook = (props) => {
+  const Facebook = (props: { facebook: { color: boolean; url: string; name: string } }) => {
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:180]: All of these "props" again... its awful to see :)

If Facebook and the otherm ethods are internal within the component, pass the "link".

File: src/molecules/Footer/Footer.tsx [hunk 18]
```diff
@@ -4,44 +4,65 @@ import { TwitterIcon } from '../../atoms/Icon/icons';
...
+}
+
+export type FooterProps = {
+  /**
+   * Array of LinkList objects
+   */
+  top: ILinkList[];
+  /**
+   * Containing data for ...
+   */
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:44]: remove comments

File: src/molecules/Footer/Footer.tsx [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+                ) : (
+                  <ArrowDownIcon style={{ height: '0.9rem' }} />
+                )}
+              </div>
+            </div>
+          </div>
+          {isActive && (
+            <div className="telia-footer__accordion-content-mobile">
+              {data.linkColumn.links.map((item) => (
+                <Item key={key++} name={item.name} url={item.url} color={item.color} />
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:84]: Passing key++, but open the <Item> code, it sets key=0?

File: src/molecules/Footer/Footer.tsx [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+            <Item key={key++} name={item.name} url={item.url} color={item.color} />
+          ))}
+        </div>
+      )}
+    </div>
+  );
+};
+
+const Item = (link: Link) => {
+  let key = 0;
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:107]: It sets key=0?

File: src/molecules/Footer/Footer.tsx [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+      <p>{data.postalCode + ' ' + data.city}</p>
+    </div>
+  );
+};
+
+const Link = (link: Link) => {
+  return (
+    <li>
+      <a className={'telia-footer__decoration' + ' ' + setColor(link.color)} href={link.url}>
+        {link.name}
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:130]: There's no "key=0" needed here, so...

File: src/molecules/Footer/Footer.tsx [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+    <>
+      {data.links.map((link) => {
+        let sosialMedia = link.name.toLowerCase();
+        if (sosialMedia.endsWith('facebook')) {
+          sosialMedia = sosialMedia + '-f';
+        }
+        const iconLookup: IconLookup = { prefix: 'fab', iconName: sosialMedia as IconName };
+        const iconDefinition: IconDefinition = findIconDefinition(iconLookup);
+        return (
+          <a className={'telia-footer__sosial' + ' ' + setColor(link.color)} key={key++} href={link.url}>
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:153]: key++, does it also increment the key variable itself

Or will this actually just set "Key=1" always?

File: src/molecules/Footer/Footer.tsx [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+            <FontAwesomeIcon icon={iconDefinition} size={'lg'} />
+            <span className={'telia-footer__sosial-media-name'}>{link.name}</span>
+            <span className={setVerticalLine(key, lenght)}></span>
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:156]: lenght > length

File: src/molecules/Footer/Footer.tsx [hunk 24]
```diff
@@ -0,0 +1,246 @@
...
+      <div>
+        <a href={data.logoColumn.url}>
+          <img src={data.logoColumn.imageUrl} />
+        </a>
+      </div>
+    );
+  }
+};
+
+const setBackgroundColor = (footerRow: FooterRow) => {
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:186]: Be consistent about GET or SET methods:

this returns backgroundColor? I would prefer "getBackgroundColor"

File: src/molecules/Footer/Footer.tsx [hunk 25]
```diff
@@ -0,0 +1,246 @@
...
+  if (!footerRow.addressColumn && !footerRow.socialLinkColumn) {
+    return 'telia-footer__purple';
+  } else {
+    return 'telia-footer__black';
+  }
+};
+
+const setColor = (color: string) => {
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:194]: getColor

File: src/molecules/Footer/Footer.tsx [hunk 26]
```diff
@@ -0,0 +1,246 @@
...
+  if (color) {
+    return 'telia-footer' + '__' + color + '-font';
+  }
+  return '';
+};
+
+const isBottomRow = (footerRow: FooterRow) => {
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:201]: is should return "true or false"

getBottomRowClass (or style)?

File: src/molecules/Footer/Footer.tsx [hunk 27]
```diff
@@ -0,0 +1,246 @@
...
+  if (footerRow.addressColumn) {
+    return 'telia-footer__container';
+  } else {
+    return 'telia-footer__set-top-horizontal-line';
+  }
+};
+
+const isBottom = (footerRow: FooterRow) => {
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:209]: Here we have another "isBottom" (not a isBottomRow..)

File: src/molecules/Footer/Footer.tsx [hunk 28]
```diff
@@ -0,0 +1,246 @@
+import React, { useState } from 'react';
+import { ArrowDownIcon, ArrowUpIcon } from '../..';
+
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { library, IconLookup, IconDefinition, findIconDefinition, IconName } from '@fortawesome/fontawesome-svg-core';
+import { fab } from '@fortawesome/free-brands-svg-icons';
```
Comment (oyvindbergbouvet) [inline src/molecules/Footer/Footer.tsx:6]: Do we need both import { fab } from '@fortawesome/free-brands-svg-icons and FontAwesome?

Not enough icons in fontawesome?

File: src/molecules/Footer/Footer.tsx (hunk 19 above)
Comment (remibengtbouvet) [reply src/molecules/Footer/Footer.tsx:84]: Changed key to a string which is best practice.

File: src/molecules/Footer/Footer.tsx (hunk 20 above)
Comment (remibengtbouvet) [reply src/molecules/Footer/Footer.tsx:107]: Changed to string which is best practice.

File: src/molecules/Footer/Footer.tsx (hunk 21 above)
Comment (remibengtbouvet) [reply src/molecules/Footer/Footer.tsx:130]: Removed since not in use.

File: src/molecules/Footer/Footer.tsx (hunk 22 above)
Comment (remibengtbouvet) [reply src/molecules/Footer/Footer.tsx:153]: Set to string according to best practice.

File: src/molecules/Footer/Footer.tsx (hunk 23 above)
Comment (remibengtbouvet) [reply src/molecules/Footer/Footer.tsx:156]: Removed one parameter since key changed to string.

File: src/molecules/Footer/Footer.tsx (hunk 24 above)
Comment (remibengtbouvet) [reply src/molecules/Footer/Footer.tsx:186]: Changed to get

File: src/molecules/Footer/Footer.tsx (hunk 25 above)
Comment (remibengtbouvet) [reply src/molecules/Footer/Footer.tsx:194]: Changed to get...

File: src/molecules/Footer/Footer.tsx (hunk 27 above)
Comment (remibengtbouvet) [reply src/molecules/Footer/Footer.tsx:209]: Removed, use isBottomRow instead.

File: src/molecules/Footer/Footer.tsx (hunk 26 above)
Comment (remibengtbouvet) [reply src/molecules/Footer/Footer.tsx:201]: Change to getBottomRowClass

File: src/molecules/Footer/Footer.tsx (hunk 28 above)
Comment (remibengtbouvet) [reply src/molecules/Footer/Footer.tsx:6]: According to documentation this is one way of solving usage of awesomefonts.

**Review by oyvindbergbouvet — APPROVED** [review]
Looks OK
- namings setBackgroundColor is a bit... but I see, it sets the style (appends) background and returns the string... so a getBackgroundClass or similary... oh well.

**Comment by remibengtbouvet** [comment]
Move from jsx to tsx Breadcrumb and Footer

**Pull request merged**